alexandria/core/views.py

```diff
@@ -168,87 +171,63 @@ def multi(self, request, **kwargs):
-            raise ValidationError(
-                _("File already has thumbnail, cannot generate multiple thumbnails.")
+    def create(self, request, *args, **kwargs):
+        serializer = self.get_serializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        if settings.ALEXANDRIA_ENABLE_AT_REST_ENCRYPTION:
+            serializer.validated_data[
+                "encryption_status"
+            ] = settings.ALEXANDRIA_ENCRYPTION_METHOD
+        obj = serializer.save()
+        if (
+            settings.ALEXANDRIA_ENABLE_THUMBNAIL_GENERATION
+            and obj.variant == models.File.Variant.ORIGINAL
+        ):
+            try:
+                create_thumbnail(obj)
+            except DjangoCoreValidationError as e:
+                raise ValidationError(*e.messages)
+        if obj.variant == models.File.Variant.THUMBNAIL:
+            try:
+                # coerce the uploaded file to a thumbnail
+                create_thumbnail(obj)
+            except DjangoCoreValidationError as e:
+                # remove the uploaded file on failure to avoid
+                # collsions
+                obj.delete()
+                raise ValidationError(*e.messages)
+            # if the original already had a thumbnail remove that
+            self.queryset.filter(original=obj.original).exclude(pk=obj.pk).delete()
+        headers = self.get_success_headers(serializer.data)
+
+        return Response(serializer.data, status=HTTP_201_CREATED, headers=headers)
+
+    @permission_classes([AllowAny])
+    @action(methods=["get"], detail=True)
+    def download(self, request, pk=None):
+        if token_sig := request.query_params.get("signature"):
+            now = timezone.now()
+            expires = float(request.query_params.get("expires"))
+            serializer_class = self.get_serializer_class()
+            host, expires, signature = serializer_class.make_signature_components(
+                pk, request, expires
             )
-
-        temp_dir, temp_filepath = get_file(file)
-        etag = create_thumbnail(file, temp_dir, temp_filepath)
-
-        if not etag:
-            return Response(
-                _("Thumbnail generation failed"), status=HTTP_400_BAD_REQUEST
+            # get expires and signature from params
+            # check if timezone.now() < expires
+            #
+            # use expires url-path, timestamp and SECRET_KEY to calculate signature
+            # compare signature from params with calculated signature
+            #
+            if now.timestamp() > expires:
+                raise PermissionDenied("Download URL expired.")
+            try:
+                assert token_sig == signature
+            except AssertionError:
+                raise PermissionDenied("Invalid signature.")
+            queryset = self.get_queryset()
+            filter_kwargs = {"pk": pk}
+            obj = get_object_or_404(queryset, **filter_kwargs)
```

Comment:
```suggestion
            obj = self.get_object()
```

Reply:
Ahh yes that absolutely makes sense. You want a fully unauthenticated request to work if you have a correctly signed URL. In that case however, I'd suggest going directly to the `File` model to avoid all DRF queryset handling

Reply:
done

Reply:
also in fdb5943